### PR TITLE
Remove still-left PATH_AVOID_DANGER_2

### PR DESCRIPTION
**Pull request by actual-nh**
#### Summary
None

#### Purpose of change

As explained in #48588, PATH_AVOID_DANGER_2 is buggy, and should be replaced by `"path_settings": { "avoid_traps": true, "avoid_sharp": true },` and PATH_AVOID_DANGER_1. There were some cases not changed by that PR (lack of spawning at the time, in general) and some introduced since then.

#### Describe the solution

I have corrected these ~~with the exceptions of 2 Exodii cyborgs that I had earlier notified @I-am-Erk about (these can readily be corrected as well if desired)~~. I have also added some max_path entries where appropriate (for feral humans and for bots with PRIORITIZE_TARGETS), including having those for all feral humans match at 30. (`max_path`, also a part of `path_settings`, is the maximum distance for which the monster will use pathing more sophisticated than movement in a straight line toward the target.)

#### Describe alternatives you've considered

Leaving supposedly-smart creatures to be vulnerable to the problems described in #48588, or to have the navigational skills of the standard zombie.

Figuring out exactly what was wrong with the code in question; while ultimately a preferable goal, this problem should be fixed in the meantime. (It may also be argued that `path_settings` should be used instead of flags anyway, although unfortunately path_settings is not entirely functional either, such as for opening doors.)

#### Testing

Loads.

#### Additional context

A thank-you to @Venera3 for spotting the problem and coming up with what to do about it (until a more-permanent fix, if any, is created).

### CODE REVIEW

File: data/json/monsters/mechsuits.json [hunk 1]
```diff
@@ -76,6 +77,7 @@
     "revert_to_itype": "broken_mech_combat",
     "death_function": { "corpse_type": "BROKEN" },
     "death_drops": { "groups": [ [ "robots", 20 ] ] },
+    "path_settings": { "max_dist": 10, "avoid_traps": true, "avoid_sharp": true },
```
Comment (actual-nh) [inline data/json/monsters/mechsuits.json:80]:
```suggestion
    "path_settings": { "max_dist": 30, "avoid_traps": true, "avoid_sharp": true },
```

File: data/json/monsters/mechsuits.json [hunk 2]
```diff
@@ -30,6 +30,7 @@
     "revert_to_itype": "broken_mech_recon",
     "death_function": { "corpse_type": "BROKEN" },
     "death_drops": { "groups": [ [ "robots", 20 ] ] },
+    "path_settings": { "max_dist": 30, "avoid_traps": true, "avoid_sharp": true },
```
Comment (actual-nh) [inline data/json/monsters/mechsuits.json:33]: Given this is a _recon_ mech, considering giving a larger max_dist.

File: data/json/monsters/mechsuits.json [hunk 3]
```diff
@@ -123,6 +125,7 @@
     "death_function": { "corpse_type": "BROKEN" },
     "special_attacks": [ { "id": "smash", "throw_strength": 96, "cooldown": 40 } ],
     "death_drops": { "groups": [ [ "robots", 20 ] ] },
+    "path_settings": { "max_dist": 10, "avoid_traps": true, "avoid_sharp": true },
```
Comment (actual-nh) [inline data/json/monsters/mechsuits.json:128]:
```suggestion
    "path_settings": { "max_dist": 30, "avoid_traps": true, "avoid_sharp": true },
```

**Comment by actual-nh** [comment]
Fixed one in DSA; ping: @ephemeralstoryteller @mlangsdorf?

File: data/mods/Dark-Skies-Above/monsters/alien_cyborgs.json [hunk 4]
```diff
@@ -318,8 +318,8 @@
     "armor_bullet": 35,
     "vision_night": 5,
     "special_attacks": [ { "type": "leap", "cooldown": 10, "max_range": 5 }, { "id": "melee_shock" } ],
-    "path_settings": { "max_dist": 5 },
+    "path_settings": { "max_dist": 5, "avoid_traps": true, "avoid_sharp": true }, },
```
Comment (actual-nh) [inline data/mods/Dark-Skies-Above/monsters/alien_cyborgs.json:321]:
```suggestion
    "path_settings": { "max_dist": 5, "avoid_traps": true, "avoid_sharp": true },
```
I fixed that already... evidently didn't update properly; sigh.

**Comment by actual-nh** [comment]
The erroring test does not seem to be related; more to inability to create target monsters due to bad positioning, looks like.

**Comment by Venera3** [comment]
`PATH_AVOID_DANGER ` is handled in `monmove.cpp` as a simple decision tree, proper pathfinding uses well, pathfinding costs and evaluation. There is a bit of a problem in that IIRC PATH_AVOID_DANGER_1 doesn't avoid dangerous fields I briefly looked into fixing, but I got distracted. Also, pathing critters can't open doors, I think that's an easy thing to adjust (the code assigns an enormously high pathing cost to opening doors for some reason).

**Comment by actual-nh** [comment]
I may take a look at the field avoidance, but it is likely to be a bit (unless I get, hrm, impulsively obsessive...).

**Comment by eltank** [comment]
> I may take a look at the field (and door) avoidance, but it is likely to be a bit (unless I get, hrm, impulsively obsessive...). Unless @eltank would like to take a look?

Disclaimer: I'm reading this code for the first time, so there may be mistakes in my analysis.

Monster navigation seems to be done by `map::route()`, which is called from `monster::move()`. `map.route()` is implemented in `pathfinding.cpp` using A*. It takes in `pathfinding_settings` which tells it the max pathfinding distance (beyond which it will fall back to unintelligent try-to-move-directly-towards-destination), whether it can open doors, use stairs and whether it should avoid traps, rough terrain or sharp terrain. There's nothing in there about field avoidance, so it will happily plan routes through fire. It does allow pathing through traps, with a large penalty.

There is field avoidance in `monster::will_move_to()` and its parent `monster::can_move_to()`. This only applies when choosing the next tile to move to (in `monster::move()`), towards either the goal (for dumb monsters) or the next tile in the (cached) path to goal (for smart monsters). Smart monsters are allowed to deviate from the path by up to 2 tiles before the cached path is invalidated, which they'll do when they can't move to a tile due to danger or obstacles.

So in the monster JSON definition there are 2 ways to control navigation behavior:
- the `path_settings` tuple, which only affects multi-step pathfinding
- the AVOID flags, which only affect single-step pathfinding

IMO the proper fix is for `map.route()` to honor the AVOID flags, which requires adding more fields to `pathfinding_settings` and setting them based on flags in `MonsterGenerator::finalize_pathfinding_settings()`. Of course, `map::route()` also needs to be changed to honor the new settings. I'm a bit busy with vehicle navigation work, but if nobody else wants to do this I'll probably get around to it eventually.

**Review by ephemeralstoryteller — APPROVED** [review]
This should be fine, so long as the drones and cyborgs still act with intelligence above that of the usual Z. Even if the 'borgs are hyped up on combat drug cocktails, they should still be able to avoid obvious tricks.

**Comment by actual-nh** [comment]
> This should be fine, so long as the drones and cyborgs still act with intelligence above that of the usual Z. Even if the 'borgs are hyped up on combat drug cocktails, they should still be able to avoid obvious tricks.

The cyborgs had "max_dist": 5, which is far below that of, say, vanilla ferals (the minimum without this PR is 10). Should this be increased? I normally was giving anything with PRIORITIZE_TARGETS a "max_dist" of 30, but left that one the same since it was already specified.

Regarding the DSA drones, which I haven't touched so far:
- The alien dispatch doesn't have anything for danger avoidance, and has "max_dist" of 5.
- Ditto for the glowdrone.
- The scidrone has PATH_AVOID_DANGER_1, without the avoidance of traps or (less significant for flying critters, I suspect) sharp things, and again a "max_dist" of 5.

Does this need correcting?

**Review by ephemeralstoryteller — COMMENTED** [review]
Increasing the pathing range sounds good, thank you!

File: data/mods/Dark-Skies-Above/monsters/alien_cyborgs.json [hunk 5]
```diff
@@ -318,8 +318,8 @@
     "armor_bullet": 35,
     "vision_night": 5,
     "special_attacks": [ { "type": "leap", "cooldown": 10, "max_range": 5 }, { "id": "melee_shock" } ],
-    "path_settings": { "max_dist": 5 },
+    "path_settings": { "max_dist": 5, "avoid_traps": true, "avoid_sharp": true },
```
Comment (actual-nh) [inline data/mods/Dark-Skies-Above/monsters/alien_cyborgs.json:321]:
```suggestion
    "path_settings": { "max_dist": 30, "avoid_traps": true, "avoid_sharp": true },
```

**Comment by satheon49** [comment]
> IMO the proper fix is for map.route() to honor the AVOID flags, which requires adding more fields to pathfinding_settings and setting them based on flags in MonsterGenerator::finalize_pathfinding_settings(). Of course, `map::route()` also needs to be changed to honor the new settings. I'm a bit busy with vehicle navigation work, but if nobody else wants to do this I'll probably get around to it eventually.

I don't agree with that conclusion for the following reasons:
1. the cached bit field pf_special is restricted in the number of flags it can support and it would need to support all field types any creature might want to path around.
2. `map::route()` has no concept for creature immunity so every field a creature can be immunte to must be defined as its own path_settings which means we would end up with avoid_fire, avoid_electric, avoid_fear, avoid_acid, avoid_sludge...
3. `monster::will_move_to()` already acts as a good gatekeeper to avoid certain tiles. It just can't be used in `map::route` directly because that method is creature agnostic.

My suggestion is the following:
Use `monster::will_move_to()` to set up a small area around the monster using `monster::will_move_to()` as a check and adding fields the monster will NOT move to to `monster::get_path_avoid()`. It does not have to be big, just a radius of 5 for example. That way monsters will have a serious look on nearby tiles and choose a path around found obstacles (but can still be lured into local "pockets" that look good before the creature realizes it has to backtrack.
```
std::set<tripoint> monster::get_path_avoid() const
{
    std::set<tripoint> ret;
    map &here = get_map();
    for( const tripoint &p : here.points_in_radius( pos(), 5 ) ) {
        if( !will_move_to( p ) ) {
            ret.insert( p );
        }
    }
    return ret;
}
```
I have not yet tested how monsters react if either the monster or player is completely surrounded by tiles it would like to avoid but even if it is a problem it could probably be solved by just giving tiles found by will_move_to a cost instead of removing them from pathing completely.

**Comment by actual-nh** [comment]
@eltank: What do you think of the below? (You know A* far better than I do, obviously!) This is from #50548:

> In a perfect world the pathfinding would use the same criteria for danger detection as monmove which would a) obsolete most of the avoid_xyz path settings and b) be probably too slow considering that the path finding algorithm has to compute this for a lot of squares.

Only if it did so non-lazily. In other words, first run A* with the current lower complexity of filtering. Then take the results as the heuristic for another iteration of A*, adding to the path cost for anything monmove would object to (although not absolutely ruling such out if at all possible). Iterate until stable. A* with a sufficiently good heuristic is quite fast, and unless a danger had not been spotted, this would be a perfect heuristic. (If a danger had been spotted, then the heuristic would be optimistic - which is required for it to work completely reliably for A*.)

**Comment by actual-nh** [comment]
> 1. the cached bit field pf_special is restricted in the number of flags it can support and it would need to support all field types any creature might want to path around.
> 2. `map::route()` has no concept for creature immunity so every field a creature can be immunte to must be defined as its own path_settings which means we would end up with avoid_fire, avoid_electric, avoid_fear, avoid_acid, avoid_sludge...
> 3. `monster::will_move_to()` already acts as a good gatekeeper to avoid certain tiles. It just can't be used in `map::route` directly because that method is creature agnostic.

Could pf_special contain an optional reference to the monster structure, so it could use that information if needed?

> My suggestion is the following:
> Use `monster::will_move_to()` to set up a small area around the monster using `monster::will_move_to()` as a check and adding fields the monster will NOT move to to `monster::get_path_avoid()`. It does not have to be big, just a radius of 5 for example. That way monsters will have a serious look on nearby tiles and choose a path around found obstacles (but can still be lured into local "pockets" that look good before the creature realizes it has to backtrack.

How will it be able to backtrack if the problem is not "known" to map::route()?

Moreover, should it really be an absolute block in most, let alone all, cases? A creature should avoid going through smoke, say, if it can, but not if that would result in taking so long to get someplace that it's an easy target in the meantime.

> 
> ```
> std::set<tripoint> monster::get_path_avoid() const
> {
>     std::set<tripoint> ret;
>     map &here = get_map();
>     for( const tripoint &p : here.points_in_radius( pos(), 5 ) ) {
>         if( !will_move_to( p ) ) {
>             ret.insert( p );
>         }
>     }
>     return ret;
> }
> ```
> 
> ~I have not yet tested how monsters react if either the monster or player is completely surrounded by tiles it would like to avoid~Monsters will stop next to an obstacle when the encounter an obstacle that they can't pass (for example a Mi-Go will stop a a wall of raging fire with length 11 compared to length 3 right now). In my opinion that makes sense for creatures concerned with self preservation. For other creatures it could probably be solved by just giving tiles found by !will_move_to a cost instead of removing them from pathing completely.

Again, I suspect that giving an increased cost - requiring passing this to map::route() in any event - will be best in most cases of a field that would not instantly destroy the monster in question. And if it's doing it over and over again for the surrounding tiles, that appears likely to be problematic in terms of performance.

There are, even with this, several uses for will_move_to not as purely something called by map::route():
- Detecting if the situation along a path has changed enough that repathing is needed (ideally with some retention of what's been figured out before);
- Prevention of stumbling movement making things too easy on the player (zombie stumbles off a cliff, etc); and
- Avoidance of damage when doing straight-line pathing due to a too-low max_dist (including the default of 0!).

**Comment by satheon49** [comment]
>How will it be able to backtrack if the problem is not "known" to map::route()?

It is known to `map::route()` because get_path_avoid() is passed to route() as a pointlist of tiles to avoid.

About the increased cost I agree, creatures should be able to path through danger if there is no other option. However this currently is not possible even if you let `map::route()` create a path through it because `monster::will_move_to()` will then block the creature from actually following this path (for example if it contains fire) and you will end up with the problem plaguing NPCs right now.

What would need to happen for your suggestion is the following:
1. will_move_to() can not block movement for creatures. It should make `map::route()` reevaluate its path but creatures would still have to follow whatever `map::route()` comes up with
2. get_path_avoid should probably be removed in this case and replaced with a creature specific weighing function that returns an additional cost for each square. This can be used as you suggested as a second pass of A* where it looks at the intermediate "good" path options:
Signature of route would look like:
```
        std::vector<tripoint> route( const tripoint &f, const tripoint &t,
                                     const pathfinding_settings &settings,
                                     const std::function<int( const tripoint & )> &opportunity_cost
        ) const;

```
And could be used with `newg += opportunity_cost(p)` somewhere appropriate during the pathfinding
A call to it would look like
```
route (..., ..., ( const tripoint & p ) {
        int cost = 0;
        // compute additional costs for a specific tile that is too expensive to cover by route during the first pass or is creature specific
        return cost;
    } )
```
Syntax might be off but the idea should be clear.
3. Field effects should have an "opportunity cost" for crossing them instead of just the dangerous boolean flag, because walking through smoke is not the same as walking though fire and even "small fire" should have a higher cost of crossing than "thick smoke". Example:
```
"id": "fd_acid",
    "type": "field_type",
    ...
    "intensity_levels": [
      { "name": "acid splatter", "sym": "5", "color": "light_green", "cost": 300
      },
      { "name": "acid streak" }, "cost": 100
      { "name": "pool of acid", "color": "green", "cost": 10 }
    ],
    "immunity_data": { "traits": [ "ACIDPROOF" ] },
    ...
```

**Pull request merged**